### PR TITLE
[ADD] base: Revoke access to your account to all devices

### PR DESCRIPTION
**Pull request by Megaaaaaa**
Description of the issue/feature this PR addresses:

No way to control sessions on other devices.

Current behavior before PR:

Prior to this, there was no way to shut an active session you didn't have the
access to anymore or close every active session remotely.

Desired behavior after PR is merged:

This system is based on the principle that every open session uses the password
hash to maintain it. When the hash is computed, salt is added to it to make it
not reversible. Thanks to that salt, it is possible to change the user's
password's hash without changing the password. Therefore, the flow of this
addition is the following :
1. User clicks on the button and uses his passord to confirm identity.
2. User's input is used to check his identity (_check_credentials()).
3. User's input is used to refresh the password hash in the db (_change_password()).
4. User is logged out.

This adds a way to close all sessions remotely and improve user's control over
their account and therefore their security.

task-3191567

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by tbe-odoo — CHANGES_REQUESTED** [review]
Hello, nothing to say regarding the python code.

Small changes to be made on the xml sides, otherwise well played :+1:

**Review by mart-e — COMMENTED** [review]
We have a model `res.users.identitycheck` that seems a bit redundant. Couldn't we use it?

File: odoo/addons/base/models/res_users.py
Comment (mart-e) [inline odoo/addons/base/models/res_users.py:1935]: that's 6min, maybe it's a bit shot for the "hmm where did I put my password again ? was it on this sheet of paper.. oh no wait" kind of scenario.

File: odoo/addons/base/security/ir.model.access.csv
Comment (mart-e) [inline odoo/addons/base/security/ir.model.access.csv:116]: we need an ir.rule as well (or use the one from res.users.identitycheck)

**Review by mart-e — COMMENTED** [review]
Could be nice to add the feature also for portal users on `/my/security`
Screenshot 2023-03-02 at 11-36-33 Odoo

File: addons/portal/controllers/portal.py
Comment (tde-banana-odoo) [reply addons/portal/controllers/portal.py:273]: And please use meaningful variables.

File: odoo/addons/base/views/res_users_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_views.xml:31]: image

Pop-up feels kinda empty without it, doesn't it ? And it won't feel logical for every user that they need to enter their password to log out from their devices. No ?

File: odoo/addons/base/views/res_users_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_views.xml:27]: image

Pop-up feels kinda empty without it, doesn't it ? And it won't feel logical for every user that they need to give their password to log out from their devices.

File: addons/portal/controllers/portal.py
Comment (Megaaaaaa) [reply addons/portal/controllers/portal.py:273]: I thought using k for keys and v for values was something known and i could use to make a understandable way to use a combination of key and value, hence the single element list.

I'll change it from Exmpl1 to Exmpl2 if you prefer :
Exmpl1:
        for k, v in [('password', pwd)]:
            if not v:
                return {'errors': {'password': {k: _("You cannot leave the password field empty.")}}}

Exmpl2:
        if not pwd:
            return {'errors': {'password': {'password': _("You cannot leave the password field empty.")}}}

File: odoo/addons/base/views/res_users_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_views.xml:27]: I thought you ment the entire "Please enter your password to confirm you wish to logout from all devices." section, sorry.
The linebreak isn't necessary indeed.

File: addons/portal/controllers/portal.py
Comment (mart-e) [inline addons/portal/controllers/portal.py:274]: `required=True` in the interface should do the same

File: addons/portal/controllers/portal.py
Comment (mart-e) [inline addons/portal/controllers/portal.py:281]:
```suggestion
                msg = _('Incrorrect password')
```
as the action failed, it's understood nothing happened

File: odoo/addons/base/models/res_users.py
Comment (mart-e) [inline odoo/addons/base/models/res_users.py:1934]:
```suggestion
        self._check_identity()
```

File: odoo/addons/base/models/res_users.py
Comment (mart-e) [inline odoo/addons/base/models/res_users.py:1930]: the `_inherit` is if you are modifying the model in another module. As you are in base, you can modify the main model definition directly.

File: addons/portal/views/portal_templates.xml
Comment (mart-e) [inline addons/portal/views/portal_templates.xml:532]:
```suggestion
                    Accesses revoked!
```

File: addons/portal/controllers/portal.py
Comment (mart-e) [inline addons/portal/controllers/portal.py:286]: functionnal question: should we be logged out in the current browser as well?

**Comment by bouvyd** [comment]
I've left a note on the task regarding a bit of UX + 2 open questions

File: addons/portal/static/src/js/portal.js
Comment (bouvyd) [inline addons/portal/static/src/js/portal.js:319]: No whitespace before ponctuation marks
```suggestion
                title: _t("Log out from all devices?"),
```

File: addons/portal/static/src/js/portal.js
Comment (bouvyd) [inline addons/portal/static/src/js/portal.js:322]:
```suggestion
                    text: _t("Log out from all devices"), classes: 'btn btn-primary',
```

People Do Not Actually Write Like This :)
(though here it matters less because right now, buttons are all-caps but that won't be the case in the new theme so please change this indeed)

File: addons/portal/static/src/xml/portal_security.xml
Comment (bouvyd) [inline addons/portal/static/src/xml/portal_security.xml:45]:
```suggestion
                You are about to log out from all devices that currently have access to this user's account.```

File: addons/portal/views/portal_templates.xml
Comment (bouvyd) [inline addons/portal/views/portal_templates.xml:532]:
```suggestion
                        Log out from all devices
```

File: odoo/addons/base/views/res_users_views.xml
Comment (bouvyd) [inline odoo/addons/base/views/res_users_views.xml:28]:
```suggestion
                        You are about to log out from all devices that currently have access to this user's account.
```

File: odoo/addons/base/views/res_users_views.xml
Comment (bouvyd) [inline odoo/addons/base/views/res_users_views.xml:546]:
```suggestion
                                    <button name="preference_revoke_all_devices" type="object" string="Log out from all devices" class="btn btn-secondary"/>
```

File: odoo/addons/base/models/res_users.py
Comment (bouvyd) [inline odoo/addons/base/models/res_users.py:945]:
```suggestion
            'name': 'Log out from all devices?',
```

**Review by bouvyd — COMMENTED** [review]
Small copy corrections

File: odoo/addons/base/views/res_users_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_views.xml:21]: id should be 'res_users_identitycheck_view_form', we try to standardize naming (model, view, type of view, ...) so that it is easy to find a given view for a given model.

File: odoo/addons/base/views/res_users_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_views.xml:546]: Note that generally action button calls method beginning with "action" (action_revoke_all_devices), just to know this is an UI business method.

File: addons/portal/static/src/js/portal.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal.js:290]: Why not adding a new file so that we avoid bloating existing files ? Moreover it helps understanding the various features and widget when looking at a module content :)

**Review by tde-banana-odoo — COMMENTED** [review]
Was pinged, saw some lights, and added some guidelines-related comments (so nothing crucial but quand même, see https://www.odoo.com/documentation/16.0/contributing/development/coding_guidelines.html if necessary :) )

Cheers !

File: addons/portal/static/src/js/portal.js
Comment (Megaaaaaa) [reply addons/portal/static/src/js/portal.js:290]: @tde-banana-odoo 
If I create a new file with the contents of my feature, maybe should I also move all the features displayed on the /my/security page so that we have a portal_security.js file instead of a portal_revoke_all_devices.js with only a few lines ?

File: addons/portal/static/src/js/portal.js
Comment (Megaaaaaa) [reply addons/portal/static/src/js/portal.js]: It has been removed :)

**Comment by Megaaaaaa** [comment]
Can i have one more review ? @tde-banana-odoo @mart-e @bouvyd

**Comment by mart-e** [comment]
@odoo/rd-framework-js could you have a look at the small JS part for portal? Thanks

File: addons/portal/static/src/js/portal.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal.js:296]: Various linting issues in this file, do you use a linter ? (spaces, missing ";", ...)

File: addons/portal/static/src/js/portal.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal.js:308]: Linting: Indent issue

File: addons/portal/static/src/js/portal.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal.js:307]: Why defining a function out of widget scope ?

File: addons/portal/static/src/js/portal.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal.js:302]: Seems a complicated JS code structure compared to what it does (a wrapper, a pointer, a function outside of widget scope, ...), unless I misunderstood something ?

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:5]: Guidelines: should be 'res_users_identitycheck_view_form', as it is a form view on a model... it helps understanding the content of a record by looking at its id (otherwise it might seem clear for you, but probably not for other people). See Odoo development guidelines for more details about naming convention :) .

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:26]: Same here, to update

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:11]: Is it necessary to have a br between h3 and a div ?

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:12]: That's uncommon to add div in standard views... we prefer to let the rendered do its job without trying to add div and custom styling, unless you have strong reasons to do so ? You may play with groups notably, depending on the display you want.

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:32]: Why so much div/br ? Can't you simplify dom ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, have some comments, mainly on styling / guidelines / view code bits. Did not test the feature itself in details as I suppose it was tested by security team and/or a PO.

Did you check with LNA by the way, as he followed various developments like that done in portal ? If not, better check with him, he is always happy to help :)

Cheers !

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_identitycheck_views.xml:5]: This entire record is something I've moved from res_users_views.xml to res_users_identitycheck_views.xml. Changing this external id would require a migration process as it is already used. It would probably be better to keep it that way.

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_identitycheck_views.xml:11]: This entire record is something I've moved from res_users_views.xml to res_users_identitycheck_views.xml. Checking the others views I put in this file was not really in the scope of my PR so I didn't want to edit it. Though, if you think it would be beneficial to take advantage of moving code to improve it, I can do that. Should I ? @tde-banana-odoo

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_identitycheck_views.xml:12]: Same here, this code doesn't belong to my and I wasn't really thinking about improving it but if you want to, I can.

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_identitycheck_views.xml:13]: Same here.

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (Megaaaaaa) [reply odoo/addons/base/views/res_users_identitycheck_views.xml:26]: As the id of the view just above is "identity_check_wizard" and shouldn't change because of a required migration process, I would think it could be better to keep the id's uniform inside the file even if it makes them a little less uniform with the rest of Odoo and therefore keep this id written like this too. This one doesn't need a migration process as it is new and therefore, if you tell me that it would be better to uniform it to Odoo rather than the id straight above, I'll rename it. Is it ?

Comment (tde-banana-odoo) [reply odoo/addons/base/views/res_users_identitycheck_views.xml:26]: Well as we target master, we may make a small imp commit indeed :) as anyway we are passing through, let us clean it quickly.

Personally I would rename all once and make them guideline compliant, as the opportunity is there as we are modifying the file. But that's not mandatory, just something I would do.

File: addons/portal/static/src/js/portal_security.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal_security.js:96]: Avoid using jquery, we plan to remove it :) use classList.add for example (Edit: I see you already use it in this file, just re-read everything once to avoid jquery addition :) )

File: odoo/addons/base/views/res_users_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_views.xml:55]: Don't forget the migration script to remove the view :) (hence red runbot by the way)

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:25]: Just to be picky: res_users_identitycheck_view_form_revokedevices (suffix is used to compare the various form view on the same model)

File: odoo/addons/base/views/res_users_identitycheck_views.xml
Comment (tde-banana-odoo) [inline odoo/addons/base/views/res_users_identitycheck_views.xml:32]: Isn't it "have access to your account" ? Seems code runs on current user, but maybe I am wrong ?

File: addons/portal/static/src/js/portal_security.js
Comment (tde-banana-odoo) [inline addons/portal/static/src/js/portal_security.js:100]: Same in this block, some remaining jquery to remove

**Review by tde-banana-odoo — COMMENTED** [review]
LGTM, some small remarks on my side :) 

Cheers and thanks !

File: addons/portal/static/src/js/portal_security.js
Comment (mart-e) [reply addons/portal/static/src/js/portal_security.js:100]: @sdegueldre could you help on what is the proper jQuery-less version of this one?

Comment (sdegueldre) [reply addons/portal/static/src/js/portal_security.js:100]: As far as I'm aware, *if* jQuery is present on the page and we don't specifically opt-out, Bootstrap dispatches its custom events using jQuery meaning a non-jQuery event listener won't be receive them, so unfortunately there is no jQ-less version of this as long as we have jQuery (unless we opt out which would require converting all event listeners)

https://getbootstrap.com/docs/5.3/getting-started/javascript/#jquery-events

**Comment by mart-e** [comment]
robodoo r+

**Pull request closed without merging**
